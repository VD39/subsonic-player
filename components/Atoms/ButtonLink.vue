<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: Component | Icon;
    iconColor?: string;
    iconPosition?: IconPosition;
    iconSize?: IconSize;
    iconWeight?: IconWeight;
    is?: ButtonLink;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconColor: undefined,
    iconPosition: 'left',
    iconSize: 'medium',
    iconWeight: 'regular',
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      'centerItems',
      $style.buttonLink,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        'centerAll fullWidth': fullWidth,
      },
    ]"
    :disabled
    draggable="false"
    :type
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="iconComponent"
      aria-hidden="true"
      :class="$style.icon"
      :color="iconColor"
      :size="ICON_SIZE[iconSize]"
      :weight="iconWeight"
    />

    <span
      ref="text"
      :class="[
        'clamp',
        $style.text,
        {
          visuallyHidden: !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.buttonLink {
  position: relative;
  gap: var(--default-space);
  padding: var(--default-space);
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  flex-shrink: 0;

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  display: block;
  margin-bottom: 1px;
}
</style>
