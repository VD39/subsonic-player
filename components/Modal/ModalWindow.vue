<script setup lang="ts">
import IconButton from '@/components/Buttons/IconButton.vue';

const { modal, closeModal } = useModal();
</script>

<template>
  <Teleport to="#teleports">
    <transition name="modal-fade">
      <div
        v-if="modal.component"
        ref="modalContainer"
        role="dialog"
        aria-modal="true"
        aria-labelledby="modal-title"
        :class="[$style.modalContainer]"
        @click.self="closeModal"
      >
        <div
          ref="modalContent"
          :class="$style.modalContent"
          role="document"
          @click.stop
        >
          <div :class="$style.header">
            <h4 v-if="modal.title" ref="title" :class="$style.title">
              {{ modal.title }}
            </h4>

            <IconButton
              icon="PhX"
              title="Close modal"
              :icon-size="32"
              @click="closeModal"
            >
              Close modal
            </IconButton>
          </div>

          <div :class="$style.body">
            <component
              :is="modal.component"
              ref="component"
              v-bind="modal.attrs"
            />
          </div>
        </div>
      </div>
    </transition>
  </Teleport>
</template>

<style module>
.modalContainer {
  @mixin align-center;

  position: fixed;
  inset: 0;
  z-index: 3;
  justify-content: center;
  width: 100%;
  height: 100%;

  &::after {
    position: fixed;
    inset: 0;
    z-index: -1;
    content: '';
    background-color: var(--black-color);
    opacity: 0.85;
  }
}

.modalContent {
  position: relative;
  width: 80%;
  background-color: var(--modal-background-color);
  border: 1px solid var(--border-color);
  border-radius: var(--border-radius-m);
  box-shadow: var(--box-shadow-medium);

  @media (--tablet-up) {
    max-width: 800px;
  }
}

.header {
  @mixin align-center;

  justify-content: space-between;
  padding: var(--space-16) var(--space-16) 0;
}

.title {
  margin-bottom: 0;
}

.body {
  padding: var(--space-16);
}

/* stylelint-disable selector-class-pattern */
:global {
  .modal-fade-enter-active,
  .modal-fade-leave-active {
    transition: opacity 0.5s ease;
  }

  .modal-fade-enter,
  .modal-fade-leave-to {
    opacity: 0;
  }
}
/* stylelint-enable selector-class-pattern */
</style>
