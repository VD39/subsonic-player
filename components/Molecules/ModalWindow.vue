<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import HeaderWithAction from '@/components/Atoms/HeaderWithAction.vue';

const { closeModal, modal } = useModal();
</script>

<template>
  <Teleport to="#teleports">
    <transition name="fade">
      <div
        v-if="modal.component"
        ref="modalContainer"
        role="dialog"
        aria-modal="true"
        aria-labelledby="modal-title"
        class="fullscreen centerAll"
        @click.self="closeModal"
      >
        <div
          ref="modalContent"
          :class="$style.modalContent"
          role="document"
          @click.stop
        >
          <HeaderWithAction :class="$style.headerWithAction">
            <h3 v-if="modal.title" ref="title" :class="$style.title">
              {{ modal.title }}
            </h3>

            <template #actions>
              <ButtonLink
                :icon="ICONS.close"
                title="Close modal"
                icon-size="large"
                icon-weight="bold"
                @click="closeModal"
              >
                Close modal
              </ButtonLink>
            </template>
          </HeaderWithAction>

          <div :class="['main', $style.body]">
            <component
              :is="modal.component"
              ref="component"
              v-bind="modal.attrs"
            />
          </div>
        </div>
      </div>
    </transition>
  </Teleport>
</template>

<style module>
.modalWindow {
  --modal-window-width-height: var(--width-height-100);

  position: fixed;
  inset: 0;
  z-index: 15;
  width: var(--modal-window-width-height);
  height: var(--modal-window-width-height);

  &::after {
    position: fixed;
    inset: 0;
    z-index: -1;
    content: '';
    background-color: var(--black-color);
    opacity: 0.85;
  }
}

.modalContent {
  position: relative;
  width: 80%;
  background-color: var(--modal-background-color);
  border: 1px solid var(--border-color);
  border-radius: var(--border-radius-medium);
  box-shadow: var(--box-shadow-medium);

  @media (--tablet-up) {
    max-width: 800px;
  }
}

.headerWithAction {
  padding: var(--space-16) var(--space-16) 0;
}

.body {
  max-height: 80vh;
  padding: var(--space-16) var(--space-16) var(--space-24);
  overflow-y: auto;
}
</style>
