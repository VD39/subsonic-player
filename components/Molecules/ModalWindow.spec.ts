import type { VueWrapper } from '@vue/test-utils';

import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { mount } from '@vue/test-utils';

import ButtonLink from '@/components/Atoms/ButtonLink.vue';

import ModalWindow from './ModalWindow.vue';

const modalMock = ref({});

mockNuxtImport('useModal', () => () => ({
  closeModal: vi.fn(() => (modalMock.value = {})),
  modal: modalMock,
}));

function factory(props = {}) {
  return mount(ModalWindow, {
    props: {
      ...props,
    },
  });
}

describe('ModalWindow', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  describe('when modal values are not set', () => {
    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('does not show the modal container element', () => {
      expect(wrapper.find({ ref: 'modalContainer' }).exists()).toBe(false);
    });
  });

  describe('when modal values are set', () => {
    beforeEach(() => {
      modalMock.value = {
        component: 'div',
      };
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the modal container element', () => {
      expect(wrapper.find({ ref: 'modalContainer' }).exists()).toBe(true);
    });

    it('shows the correct component', () => {
      expect(
        (wrapper.find({ ref: 'component' }).element as HTMLElement).tagName,
      ).toBe('DIV');
    });

    describe('when title is not set', () => {
      it('does not show the title element', () => {
        expect(wrapper.find({ ref: 'title' }).exists()).toBe(false);
      });
    });

    describe('when title is set', () => {
      beforeEach(() => {
        modalMock.value = {
          component: 'div',
          title: 'Title',
        };
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('shows the title element', () => {
        expect(wrapper.find({ ref: 'title' }).exists()).toBe(true);
      });
    });

    describe('when modal content is clicked', () => {
      beforeEach(async () => {
        await wrapper.find({ ref: 'modalContent' }).trigger('click');
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('does not remove the component', () => {
        expect(wrapper.find({ ref: 'modalContainer' }).exists()).toBe(true);
      });
    });

    describe('when modal container is clicked', () => {
      beforeEach(async () => {
        await wrapper.find({ ref: 'modalContainer' }).trigger('click');
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('removes the component', () => {
        expect(wrapper.find({ ref: 'modalContainer' }).exists()).toBe(false);
      });
    });

    describe('when the ButtonLink component is clicked', () => {
      beforeEach(async () => {
        wrapper.findComponent(ButtonLink).vm.$emit('click');
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('removes the component', () => {
        expect(wrapper.find({ ref: 'modalContainer' }).exists()).toBe(false);
      });
    });
  });
});
