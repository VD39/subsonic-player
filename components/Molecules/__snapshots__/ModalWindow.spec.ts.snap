// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModalWindow > when modal values are not set > matches the snapshot 1`] = `
"<teleport-stub to="#teleports">
  <!--v-if-->
</teleport-stub>"
`;

exports[`ModalWindow > when modal values are set > matches the snapshot 1`] = `
"<teleport-stub to="#teleports">
  <div aria-labelledby="modal-title" aria-modal="true" class="fullscreen centerAll fade-enter-from fade-enter-active" role="dialog">
    <div class="modalContent" role="document">
      <div class="centerItems spaceBetween headerWithAction">
        <div class="centerItems"><button class="centerItems buttonLink" draggable="false" type="button" title="Close modal">
            <phx aria-hidden="true" class="icon" size="32" weight="bold"></phx><span class="clamp text visuallyHidden"> Close modal </span>
          </button></div>
      </div>
      <div class="main body">
        <div></div>
      </div>
    </div>
  </div>
</teleport-stub>"
`;

exports[`ModalWindow > when modal values are set > when modal container is clicked > matches the snapshot 1`] = `
"<teleport-stub to="#teleports">
  <div aria-labelledby="modal-title" aria-modal="true" class="fullscreen centerAll fade-leave-from fade-leave-active" role="dialog">
    <div class="modalContent" role="document">
      <div class="centerItems spaceBetween headerWithAction">
        <div class="centerItems"><button class="centerItems buttonLink" draggable="false" type="button" title="Close modal">
            <phx aria-hidden="true" class="icon" size="32" weight="bold"></phx><span class="clamp text visuallyHidden"> Close modal </span>
          </button></div>
      </div>
      <div class="main body">
        <div></div>
      </div>
    </div>
  </div>
  <!--v-if-->
</teleport-stub>"
`;

exports[`ModalWindow > when modal values are set > when modal content is clicked > matches the snapshot 1`] = `
"<teleport-stub to="#teleports">
  <div aria-labelledby="modal-title" aria-modal="true" class="fullscreen centerAll" role="dialog">
    <div class="modalContent" role="document">
      <div class="centerItems spaceBetween headerWithAction">
        <div class="centerItems"><button class="centerItems buttonLink" draggable="false" type="button" title="Close modal">
            <phx aria-hidden="true" class="icon" size="32" weight="bold"></phx><span class="clamp text visuallyHidden"> Close modal </span>
          </button></div>
      </div>
      <div class="main body">
        <div></div>
      </div>
    </div>
  </div>
</teleport-stub>"
`;

exports[`ModalWindow > when modal values are set > when the ButtonLink component is clicked > matches the snapshot 1`] = `
"<teleport-stub to="#teleports">
  <div aria-labelledby="modal-title" aria-modal="true" class="fullscreen centerAll fade-enter-from fade-leave-from fade-leave-active" role="dialog">
    <div class="modalContent" role="document">
      <div class="centerItems spaceBetween headerWithAction">
        <div class="centerItems"><button class="centerItems buttonLink" draggable="false" type="button" title="Close modal">
            <phx aria-hidden="true" class="icon" size="32" weight="bold"></phx><span class="clamp text visuallyHidden"> Close modal </span>
          </button></div>
      </div>
      <div class="main body">
        <div></div>
      </div>
    </div>
  </div>
  <!--v-if-->
</teleport-stub>"
`;

exports[`ModalWindow > when modal values are set > when title is set > matches the snapshot 1`] = `
"<teleport-stub to="#teleports">
  <div aria-labelledby="modal-title" aria-modal="true" class="fullscreen centerAll" role="dialog">
    <div class="modalContent" role="document">
      <div class="centerItems spaceBetween headerWithAction">
        <h3 class="title">Title</h3>
        <div class="centerItems"><button class="centerItems buttonLink" draggable="false" type="button" title="Close modal">
            <phx aria-hidden="true" class="icon" size="32" weight="bold"></phx><span class="clamp text visuallyHidden"> Close modal </span>
          </button></div>
      </div>
      <div class="main body">
        <div></div>
      </div>
    </div>
  </div>
</teleport-stub>"
`;
